Sql2o removal for MySQL JDBC

mashurex opened the pull request:
@picaron This is a sample of the `java.sql.*` only implementation from #424 , I've included the `MetadataDAO` implementation- all the new work is in the `com.netflix.conductor.dao.jdbc` package so that it's easier to compare side by side with the existing `com.netflix.conductor.dao.mysql` package.

# Approach
Maintain the same developer flow and ergonomics that @picaron set with the usage of the Sql2o library and `MySQLBaseDAO`.

I essentially reused all existing queries, just swapping out named params for standard `PreparedStatement` bind question marks, `?`.

I created a few `@FunctionalInterface` instances for Java 8 style `Consumer` and `Function` patterns that allow for `SQLExceptions` to be handled by wrapper methods in the `JdbcBaseDAO` and `Query` classes.

## JdbcBaseDAO Class
The `jdbc.JdbcBaseDAO` mimics the `MySQLBaseDAO` conventions heavily with the `getWithTransaction` and `withTransaction` methods, but adds additional convenience wrappers as well. 

## Query Builder Class
the `jdbc.Query` class is inspired by the nice ergonomics of Sql2o, bet eschews a lot of the generic conversion and data type handling for the more simple use cases encountered in the Conductor DAOs.

**Highlights**

- Builder pattern _addParameter_ methods for quicker chaining of parameter binding.
- Primitive batch parameter addition via the `addParameters(Object...)` method
- `addJsonParameter(Object)` as a convenience method for the repeated persistence of Conductor objects as JSON strings in columns
- `SQLException` conversion to runtime `ApplicationException` exceptions to remove the massive `try/catch` boilerplate required by `PreparedStatement` and `ResultSet` interactions
- Convenience methods like `boolean exists()` and `boolean delete()` that eschew the boiler plate code of checking a `ResultSet` for affected rows or retrieving a scalar value
- Use of Java _try with resources_ for automatic management of `PreparedStatement` and `ResultSet` instances

picaron commented:
sounds very promising @mashurex !

It would be good to get approval from @v1r3n or another of the project owners before you spend more time on the PR. If they have no objection with this change I certainly don't.

Also I forgot to mention one of the reason for choosing Sql2o over other libraries was performance but raw JDBC should be even faster 🏎 

I love the idea of removing the Sql2o dependency but keeping the same "ergonomics" as you say 👍

mashurex commented:
@picaron It only took about 30 seconds to guess why Sql2o would be included. Raw JDBC in Java is painfully repetitive. Getting more familiar w/ Sql2o, I was able to come up with what I feel like is a decent wrapper around the native calls without introducing a crazy amount of complexity or re-inventing the wheel and yet introduce some of the very nice things that Sql2o gives to a developer.

picaron commented:
I haven't used Raw JDBC in a long time but I remembered it was very painful because it required so much repetitive boilerplate code.

a decent wrapper with an API similar to Sql2o is the way to go I think because otherwise if the DAOs end-up being polluted with boilerplate then they become much harder to read and maintain ...

mashurex commented:
@v1r3n and @picaron this PR is code complete from a DAO interface implementation standpoint. I did not put in a `Module` as I'm waiting for your thought on this work. If taken, I'm assuming we'd move this into and replace the `mysql` package.

- ExecutionDAO implemented
- MetadataDAO implemented
- QueueDAO implemented
- All `MySQL*Test` unit tests re-implemented for the Jdbc based DAOs
- Brought in and ported the work from #399 (as well as tests)

v1r3n commented:
@mashurex @picaron Thanks for the PR and details on the JDBC vs. Sql2o
IMHO, given that we do not have a lot of complexity in the DAOs and also the nature of the DAOs do not evolve that frequently as otherwise in a typical app, plain JDBC with some wrappers to make things easier will be a good balance.  

Using direct JDBC also helps those who are not very familiar with Sql2o or other ORM frameworks get up to speed quickly.  Conductor is a framework in the end and given that I would prefer to keep it as light as possible.

My +1 is for JDBC.

v1r3n commented:
@mashurex do you want to add the rest of the stuff (module etc) and update the PR so it replaces the existing implementation?

@picaron @pctreddy

mashurex commented:
@v1r3n Will do, was just waiting on everyone's feedback.

mashurex commented:
@v1r3n and @picaron, the work is ready for review and merge, pending any feedback from you guys.

picaron commented on mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLBaseDAO.java:
```diff
@@ -1,91 +1,202 @@
+ */
+public abstract class MySQLBaseDAO {
+    private static final List<String> EXCLUDED_STACKTRACE_CLASS = ImmutableList.of(
+        "com.netflix.conductor.dao.jdbc.JdbcBaseDAO", "java.lang.Thread");
```
this probably needs to be updated

mashurex replied:
Oops! Good catch!

mashurex commented:
@picaron I fixed that stack trace class string.

picaron commented:
thanks @mashurex , didn't have time to do a full review

I'll continue asap

picaron commented on mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLBaseDAO.java:
```diff
@@ -1,91 +1,202 @@
+ */
+public abstract class MySQLBaseDAO {
+    private static final List<String> EXCLUDED_STACKTRACE_CLASS = ImmutableList.of(
+        "com.netflix.conductor.dao.mysql.MySQLBaseDAO", "java.lang.Thread");
```
maybe you could replace the strings with the class reference, for example:

`com.netflix.conductor.dao.mysql.MySQLBaseDAO.class.getName()` and `java.lang.Thread.class.getName()`

this way if those classes are renamed will make sure we don't forget to update this line

picaron commented on mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLBaseDAO.java:
```diff
@@ -1,91 +1,202 @@
+     * @return The result of {@code TransactionalFunction#apply(Connection)}
+     * @throws ApplicationException If any errors occur.
+     */
+    protected <R> R getWithTransaction(TransactionalFunction<R> function) {
```
I love what you did with the overall PR but this method I find it very hard to read

it's indeed difficult to manage all the flows of exceptions. The `Connection` object supports auto-closeable so it can be used in a try with resource block...

I tried to simplify the code a bit to improve readability, here's what I came up with:

```
protected <R> R getWithTransaction(TransactionalFunction<R> function) {
	Instant start = Instant.now();
	StackTraceElement caller = getCaller();
	logger.trace("{} : starting transaction", caller.getMethodName());

	try(Connection tx = dataSource.getConnection()) {
		boolean previousAutoCommitMode = tx.getAutoCommit();
		tx.setAutoCommit(false);
		try {
			R result = function.apply(tx);
			tx.commit();
			return result;
		} catch (SQLException ex) {
			tx.rollback();
			throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
		} finally {
			tx.setAutoCommit(previousAutoCommitMode);
		}
	} catch (SQLException ex) {
		throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
	} finally {
		logger.trace("{} : took {}ms", caller.getMethodName(), Duration.between(start, Instant.now()).toMillis());
	}
}
```

mashurex replied:
This is much easier to read, the reason the checks for `logger.isTraceEnabled()` were being used is that this method gets called very frequently and I didn't want to burden the execution with parsing StackTraceElements if they weren't going to be used. 

I didn't spend a lot of time measuring what kind of execution cost `getCaller()` has, but I figured that if we are not logging `trace` level messages, then there is no reason to parse the `StackTraceElement`. This adds a lot of noise with if statements, but it gives us the best _likelihood_ of faster execution times.

I'm really not sure why I didn't use a try with resources on the connection, it was probably an oversight on my part. I also see that you added `previousAutoCommitMode` which is a fantastic idea.

picaron replied:
regarding the logger, since here we are using SLF4J with parameterized logging, the logger.isTraceEnabled() is redundant.

have a look at : https://www.slf4j.org/faq.html#logging_performance

mashurex replied:
That's correct and we're using it, except for we are executing `getCaller()` no matter if the Logger is trace enabled or not. In my experience with this there's a trade off between a little harder to read and not executing debug/trace level calculations if we don't need to. In my other projects we usually have utility classes to handle this kind of stuff to reduce the boilerplate.

Again, I haven't measured the performance impact in any real detail but I assume that wrapping the `getCaller()` in a `isTraceEnabled()` block is more of a micro-optimization than a true performance benefit. I agree that it makes it harder to follow the code, so I think it's perfectly reasonable to follow your approach of removing the guard blocks and simplifying it.

picaron replied:
you are right @mashurex I overlooked that!

how about this: 

```
protected <R> R getWithTransaction(TransactionalFunction<R> function) {
        Instant start = Instant.now();
        LazyToString callingMethod = getLazyCallingMethod();
        logger.trace("{} : starting transaction", callingMethod);

        try(Connection tx = dataSource.getConnection()) {
            boolean previousAutoCommitMode = tx.getAutoCommit();
            tx.setAutoCommit(false);
            try {
                R result = function.apply(tx);
                tx.commit();
                return result;
            } catch (Throwable th) {
                tx.rollback();
                throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, th);
            } finally {
                tx.setAutoCommit(previousAutoCommitMode);
            }
        } catch (SQLException ex) {
            throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
        } finally {
            logger.trace("{} : took {}ms", callingMethod, Duration.between(start, Instant.now()).toMillis());
        }
    }

    protected final LazyToString getLazyCallingMethod() {
        return new LazyToString(() -> Arrays.stream(Thread.currentThread().getStackTrace())
            .filter(ste -> !EXCLUDED_STACKTRACE_CLASS.contains(ste.getClassName()))
            .findFirst()
            .map(StackTraceElement::getMethodName)
            .orElseThrow(() -> new NullPointerException("Cannot find Caller")));
    }
```

and the `LazyToString` class:

```
package com.netflix.conductor.dao.mysql;

import java.util.function.Supplier;

class LazyToString {
    private final Supplier<String> supplier;

    LazyToString(Supplier<String> supplier) {
        this.supplier = supplier;
    }

    @Override
    public String toString() {
        return supplier.get();
    }
}
```

mashurex replied:
I really like this. I'll get it in today.

mashurex commented:
@picaron I've added the fix from #448 into this PR as well as the lazy stack trace retrieval.

picaron commented:
good work @mashurex ! thanks!

there also seems to be a merge conflict to resolve...

mashurex commented:
@picaron This is probably where my Github PR naïveté comes in as I wasn't exactly sure who is responsible for merging the conflict. I'll get it fixed up.

mashurex commented:
@picaron we’re merge ready but I’d like your opinion on the unit tests I added to the `MySQLQueueDAOTest` test class.

I was having some timing/sequence failures when the build would push to Travis and run the unit tests. I _think_ it’s because the timestamps in the DB columns don’t have MS resolution and the VM layer in Travis when calling `System.currentTimeMillis()` is only giving 10ms resolution. 

The resulting issue is that if we push X messages onto the queue in a for loop, they all have the exact same `created_on` (and `deliver_on` for a given offset). That means when we run our peek query against the DB we have no guarantee of the exact order of messages that hit the queue in the same second with the same deliver on, because they all have the same timestamp for any given 1 second window.

I changed my tests from expecting a particular order to testing for the messages getting polled being in the collection of expected messages.

mashurex commented:
@picaron and @v1r3n is there any chance we can get this reviewed and merged soon? Keeping it up to date and conflict free is getting a bit troublesome.

mashurex commented:
Sounds good. I will get the migrations fixed ASAP.
On Sat, Apr 14, 2018 at 09:24 Pascal Chouinard <notifications@github.com>
wrote:

> @mashurex <https://github.com/mashurex> I agree, let's get this done
> asap. I think we are very close.
>
> Latest code looks good to me overall and all tests run fine but there is
> one remaining issue: the existing flyway migration script was modified so
> everyone upgrading will be getting a flyway conflict.
>
> Could you please put the changes in a new V2__xyz.sql migration file? You
> can for example add a V2__update_queue_message.sql file with an alter
> table statement in it.
>
> Ideally, I'd like to test running my workflows using your PR to make sure
> everything still works before we proceed with the merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/conductor/pull/426#issuecomment-381340988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABN59FT5lqvMoonyo4vqhG9Iupfd0AbIks5toiKlgaJpZM4SEVRH>
> .
>

mashurex commented:
@picaron Schema changes moved to V2 and pushed.

picaron commented on mysql-persistence/src/test/resources/logback-test.xml:
```diff
@@ -1,17 +1,18 @@
-	<logger name="org.flywaydb" level="WARN" />
-	<logger name="com.zaxxer.hikari" level="WARN" />
-	<logger name="ch.vorburger" level="WARN" />
+  <logger name="org.sql2o" level="INFO"/>
```
this can probably be removed

picaron commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,5 @@
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-bin.zip
```
maybe this file should be left as is was before?

picaron commented:
@mashurex is it possible to squash your commits into a single one? 

once merged into the target repo, it would keep the commits history a bit cleaner

picaron commented:
@v1r3n are ok with merging this?

mashurex commented:
@picaron I squashed the changes to make the history cleaner.

picaron commented:
@mashurex btw, I don't see the sql2o dependency was removed from the `build.gradle`, that should be removed

you can also remove `  <logger name="org.sql2o" level="INFO"/>` from `logback_test.xml`

picaron commented:
@v1r3n I'm not familiar with the `dependencies.lock` how can that be updated? is there a command or we can just remove all references to sql2o from there?

mashurex commented:
@picaron and @v1r3n I removed sql2o references and updated the dependencies.lock as well. Travis builds 1018 and 1019 seem to fail in an unrelated test in the same way that build 1017 did which is from PR #514- so I think there's something currently broken on the dev branch.

picaron commented:
seems @mashurex has address all the latest comments

good to merge @v1r3n ?
